### PR TITLE
Htn devel

### PR DESCRIPTION
**Pull request by oscar-lima**
HTN Planning with CHIMP planner in ROSPlan

### CODE REVIEW

File: rosplan_planning_system/test/hierarchical_domains/chimp/restaurant/serve_coffee_problem.pdl [hunk 1]
```diff
@@ -0,0 +1,104 @@
+(Problem
```
Comment (m312z) [inline rosplan_planning_system/test/hierarchical_domains/chimp/restaurant/serve_coffee_problem.pdl:1]: Should this file be "pdl" file extension?

Comment (m312z) [reply rosplan_planning_system/test/hierarchical_domains/chimp/restaurant/serve_coffee_problem.pdl:1]: Same question applies to the other files.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -42,6 +42,11 @@ jobs:
         shell: bash
       - name: Build workspace
         run: catkin build --summarize --no-status
+      - name: Build CHIMP
```
Comment (m312z) [inline .github/workflows/test.yml:45]: Should building CHIMP be a part of building ROSPlan? No other planners are compiled as a part of compiling ROSPlan - indeed, it would be too much to ask a user of ROSPlan to resolve all dependencies of and compile all planners, even though they would normally only be interested in using one.

Comment (gerardcanal) [reply .github/workflows/test.yml:45]: Why isn't it built as part of rosplan dependencies as it's defined as one? Rddlsim for instance is built that way during catkin_make.

File: rosplan_planning_system/test/hierarchical_domains/chimp/restaurant/race_waiter_domain.ddl [hunk 3]
```diff
@@ -0,0 +1,724 @@
+##################
```
[the remark this answers is not in the document]
Comment (sebastianstock) [reply rosplan_planning_system/test/hierarchical_domains/chimp/restaurant/race_waiter_domain.ddl:1]: The domain description is in CHIMP's own format, which is why "ddl" is correct.

File: rosplan_planning_system/test/hierarchical_domains/chimp/restaurant/serve_coffee_problem.pdl (hunk 1 above)
Comment (sebastianstock) [reply rosplan_planning_system/test/hierarchical_domains/chimp/restaurant/serve_coffee_problem.pdl:1]: Same as  for the ddl also the problem file is in CHIMP's pdl-format.

Comment (m312z) [reply rosplan_planning_system/test/hierarchical_domains/chimp/restaurant/serve_coffee_problem.pdl:1]: I suspected so - just wanted to double-check.

File: .github/workflows/test.yml (hunk 2 above)
Comment (sebastianstock) [reply .github/workflows/test.yml:45]: I am actually building it here is so that it does not have to be compiled by all users:
Because CHIMP uses gradle it normally does not need to be compiled in advance.
However, the first time you call CHIMP via gradle it is significantly slower (especially with the github action) such that the tests failed due to a timeout. Therefore, for the CI I am building it in advance such that gradle downloads dependencies and compiles chimp - making the subsequent tests pass.
A normal user does not have to build CHIMP - especially not if he does not want to use it.

If you prefer to build it during catkin_make I can try to put it there.

Comment (gerardcanal) [reply .github/workflows/test.yml:45]: Ok, I see this is only for the test. So it is not being built by default?
I understand CHIMP only needs to build only once, while the workspace is building. There should not be a timeout there during the catkin build. I'm not familiar with gradle so I am not sure about. 
Maybe we could add a CMake flag to compile CHIMP or not, so we can specify whether to build or not during catkin build? And then we set it for the GitHub actions. This may be easier for users than having to build chimp manually and compiling two things.

Comment (sebastianstock) [reply .github/workflows/test.yml:45]: The timeout does not happen in the catkin build, but in the first test of the planner interface for CHIMP. It is part of the planner command that the planner inferface uses to call CHIMP: `timeout 50 " + chimp_path + "/gradlew run -p " + chimp_path + " -Dexec.args=\"DOMAIN PROBLEM -o=OUTPUT\"`
This only fails on the slow CI server.

Your suggestion to build it dependent on a flag with catkin sounds good and it might be the cleaner solution for this.

Comment (sebastianstock) [reply .github/workflows/test.yml]: With 2001212 CHIMP is built by catkin if the cmake variable `ROSPLAN_COMPILE_CHIMP` is set to `TRUE` as suggested by @gerardcanal.

Comment (gerardcanal) [reply .github/workflows/test.yml]: This looks great Sebastian! I think we should probably do the same with RDDLSim and the other things that compile there. I'll open an issue to remember this.

File: rosplan_planning_system/src/ProblemGeneration/ProblemGeneratorFactory.cpp [hunk 4]
```diff
@@ -10,6 +10,7 @@ namespace KCL_rosplan {
     ProblemGeneratorPtr ProblemGeneratorFactory::createProblemGenerator(ProblemGeneratorFactory::ProbGen pg_type, const std::string& kb) {
         if (pg_type == ProbGen::PDDL) return ProblemGeneratorPtr(new PDDLProblemGenerator(kb));
         else if (pg_type == ProbGen::RDDL) return ProblemGeneratorPtr(new RDDLProblemGenerator(kb));
+        else if (pg_type == ProbGen::CHIMP) return ProblemGeneratorPtr(new CHIMPProblemGenerator(kb));
```
Comment (gerardcanal) [inline rosplan_planning_system/src/ProblemGeneration/ProblemGeneratorFactory.cpp:13]: Why is CHIMP here but HDDL in the KB? 
Same with the problem generator. Does CHIMP have its own problem format which is different from HDDL?

Comment (sebastianstock) [reply rosplan_planning_system/src/ProblemGeneration/ProblemGeneratorFactory.cpp:13]: Yes, exactly. CHIMP has its own format for problems and domains.

Comment (gerardcanal) [reply rosplan_planning_system/src/ProblemGeneration/ProblemGeneratorFactory.cpp:13]: And is this format HDDL then? My main question was the difference in names for the KB and the PG, I think this should be consistent (if the KB and the PG use the same kind of files).

Comment (oscar-lima) [reply rosplan_planning_system/src/ProblemGeneration/ProblemGeneratorFactory.cpp:13]: No, CHIMP extensions are ".ddl" for the domain and ".pdl" for the problem. However in future we want to support HDDL planners, so for now we are defining 2 domains, 1 in HDDL, 1 in CHIMP format, KB loads the domain in HDDL format and the chimp problem generator reads from KB and generates a problem in CHIMP format. In future HDDL problem generators are expected, and in general full support for HDDL planners.

Comment (m312z) [reply rosplan_planning_system/src/ProblemGeneration/ProblemGeneratorFactory.cpp:13]: Gerard mentioned: it would be nice to have an alias CHIMP KB, which is just the HDDL knowledge base. This would be clearer overall. We can put this on a todo list?

File: rosplan_planning_system/scripts/direct_esterel_parser.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env python
```
Comment (m312z) [inline rosplan_planning_system/scripts/direct_esterel_parser.py:1]: This is a new node in ROSPlan. Should it be a part of this PR? It looks as though it is used for testing, in which case it should either be removed, or moved to a separate PR in rosplan_demos.

File: rosplan_planning_system/launch/includes/chimp_planner_interface.launch [hunk 6]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0"?>
```
Comment (m312z) [inline rosplan_planning_system/launch/includes/chimp_planner_interface.launch:1]: There is no PANDA planner launch - does it use the same as CHIMP?

File: rosplan_planning_system/scripts/direct_esterel_parser.py (hunk 5 above)
Comment (sebastianstock) [reply rosplan_planning_system/scripts/direct_esterel_parser.py:1]: Yes, it should be part of the pull request. 
As CHIMP already has an internal representation in form of a temporal constraint network, I added an option to CHIMP such that it can directly write the plan's Esterel representation to a file.
This node just converts this output into a `EsterelPlan` ROS message and publishes it.

Comment (m312z) [reply rosplan_planning_system/scripts/direct_esterel_parser.py:1]: That makes sense. Might be worth editing the comment on line 12 to specify that it means ESTEREL plan YAML as string. Planner output is normally something else.

File: rosplan_planning_system/launch/includes/chimp_planner_interface.launch (hunk 6 above)
Comment (oscar-lima) [reply rosplan_planning_system/launch/includes/chimp_planner_interface.launch:1]: The integration with PANDA and in general with HDDL planners is still work under progress. At least Lilotane and PANDA are planned. The KB already accepts HDDL formalism which is a step in the right direction.

File: rosplan_planning_system/scripts/direct_esterel_parser.py (hunk 5 above)
Comment (sebastianstock) [reply rosplan_planning_system/scripts/direct_esterel_parser.py:1]: You are right. I have updated the comment.
